csrc/scheduler/resize.cpp

```diff
@@ -0,0 +1,194 @@
+      // ID only. For example,
+      //
+      // t0: [i0]
+      // t1: [b1]
+      // t2 = t0 + t1
+      // t3 = slice(t2)
+      //
+      // Then, propagating the slice to its inputs would try to
+      // propagate the resize op to b1 as well, which would fail due
+      // to issue #3571
+      const auto& input_group = broadcast_graph.toGroup(resize->in());
+      if (std::any_of(
+              input_group->begin(), input_group->end(), [](Val* inp_val) {
+                return inp_val->as<IterDomain>()->isBroadcast();
+              })) {
+        scheduler_debug_utils::canScheduleRejectReason(
+            schedulerType(), "Resize of a broadcast ID is not allowed.");
+        return false;
+      }
+    }
+  }
+
+  // This doesn't work yet due to issue #3571
+  auto ref_tv = getReferenceTensor(fusion);
+  if (std::any_of(
+          ref_tv->getLogicalDomain().begin(),
+          ref_tv->getLogicalDomain().end(),
+          [](IterDomain* logical_id) { return logical_id->isBroadcast(); })) {
+    return false;
+  }
+
+  // Disable the scheduler if there's a squeeze op. The loop option
+  // may also need to be enabled in that case, but that option is not
+  // turned on automatically yet.
+  if (ir_utils::hasOpsOfType<SqueezeOp>(fusion)) {
+    return false;
+  }
+
+  return true;
+}
+
+std::unique_ptr<HeuristicParams> ResizeScheduler::computeHeuristics(
+    Fusion* fusion,
+    SchedulerRuntimeInfo& runtime_info,
+    HeuristicDataCache* data_cache) {
+  FUSER_PERF_SCOPE("ResizeScheduler::computeHeuristics");
+  auto params = std::make_unique<HeuristicParams>(SchedulerType::Resize);
+  params->cparams.index_type = runtime_info.getIndexType();
+  return params;
+}
+
+void ResizeScheduler::schedule(Fusion* fusion, const HeuristicParams* params) {
+  FUSER_PERF_SCOPE("ResizeScheduler::schedule");
+
+  FusionGuard fg(fusion);
+
+  scheduler_utils::clearMemorySpace(fusion);
+
+  scheduler_utils::cacheInputs(fusion, true);
+  scheduler_utils::cacheAndForkOutputs(fusion, true);
+
+  for (auto expr : fusion->exprs()) {
+    if (!expr->isOneOf<SliceOp, PadOp>()) {
+      continue;
+    }
+
+    scheduler_tools::propagateResizeToInputs(expr);
+  }
+
+  auto ref_tv = getReferenceTensor(fusion);
+
+  // Just simple scheduling for now.
+  // TODO: Do something smarter. Can just use the pointwise scheduler?
+
+  // Make sure the DID ID located at the outermost position
+  int64_t reorder_pos = 0;
+  std::unordered_map<int64_t, int64_t> old2new;
+  for (const auto i : c10::irange(ref_tv->getLoopDomain().size())) {
+    if (isParallelTypeDeviceDim(ref_tv->axis((int64_t)i)->getParallelType())) {
+      old2new.emplace((int64_t)i, reorder_pos);
+      ++reorder_pos;
+    }
+  }
+  ref_tv->reorder(old2new);
+
+  // Schedule only the remaining IDs
+  const auto outermost_pos = (int64_t)old2new.size();
+  ref_tv->flatten(outermost_pos);
+  ref_tv->split(outermost_pos, 128);
+  ref_tv->split(outermost_pos, 1 << 14);
```

Comment:
This is a little off-topic but is there any particular reason to split by 16K here? In the pointwise scheduler I think the split is into 64K blocks. Would it be better to just make this a heuristic param so that it could be set to some multiple of the number of SMs?

Reply:
Nothing particular. This is just a random simple scheduling I put here for now. These heuristics parameters would need to go through the tuning process once all the building blocks are in place (much like the matmul scheduler development).